### ISSUE TITLE
How to write "0" value?

### ISSUE THREAD

**Issue by cbagpipe** [OlegZee/NEbml#10]
When I'm trying to write cue point with following code
```csharp
cuePoint.Write(VInt.FromEncoded(0xB3), absoluteTimestamp); /* CueTime */
```
and `absoluteTimestamp` is `0`. I'm getting `0xB3 0x80` in raw bytes instead of expected `0xB3 0x81 0x00`
Am I doing something wrong?

**Comment by OlegZee** [comment]
No, you are doing it right.
According to https://www.rfc-editor.org/rfc/rfc8794.html#name-vint-examples the value of 0 is encoded with 1 octet:
1xxx xxxx where xxxxxxx = 0 results in 0x80.

**Comment by andrekoehler** [comment]
I also stumbled upon this issue. The matroska element "Tracks.TrackEntry.FlagLacing" has a defined default value of 1 (see https://matroska.org/technical/elements.html). I want it to have a value of 0. In my old MKV files created with an older version of NEbml I can see it being correctly written as **9C 81 00**. With the new version 0.10.1 of the NEbml NuGet package I get **9C 80** instead. https://www.rfc-editor.org/rfc/rfc8794.html#section-7.2 states "_If the EBML Element is not defined to have a default value, then an Unsigned Integer Element with a zero-octet length represents an integer value of zero._". So in my opinion NEbml incorrectly assumes that the default value for integers is always 0, but this is not true.

**Comment by OlegZee** [comment]
Well, section 6.1 (https://www.rfc-editor.org/rfc/rfc8794.html#section-6.1) is even more interesting.
If I understand it correctly long/ulong values should use length of 1 or more.
I could provide WriteDefault(VInt elementId) for legacy behavior (zero length), which could be used for strings as well.
Btw empty string will be encoded with 0 length as well.

Would it work?

**Comment by andrekoehler** [comment]
May you could just add an overload with nullable values to the EbmlWriter.Write method and add EbmlReader.ReadNullable* methods?
Examples:
```
// A value of null will be written with a VINT length of zero bytes.
public int Write(VInt elementId, long? value)
public int Write(VInt elementId, ulong? value)

// A VINT length of zero will be returned as null
public long? ReadNullableInt()
public ulong? ReadNullableUInt()
```

This way during writing I could write an empty integer by providing a typed null (with the expression `default(long?)`) or write a one-byte value of zero using `(long?)0`.

I believe old would not be affected because it would always call the Write-method-overload which does not use the System.Nullable type.

**Comment by OlegZee** [comment]
I suppose it's redundant.
Which one would be used if I write code like this `Write(DurationElementId, null)`?
Besides that we don't care about backword compabitility so both options look the same.

Regarding reading: looks fine to me.

**Comment by andrekoehler** [comment]
> I suppose it's redundant.
> Which one would be used if I write code like this `Write(DurationElementId, null)`?
Good point, I did not think about that. I just tested it with a test project and it seems like there is a priority between integer types when selecting between Nullable overloads. When having to chose between int?, uint?, long? and ulong? C# picks the overload with int? but as soon as an overload with string exists I get "error CS0121: The call is ambiguous". So my idea was not good I guess.

Next idea: Specify the default value in the write call. If the value equals the default value, the element is written with a length of zero:
```
public int Write(VInt elementId, long value, long defaultValue = 0)
public int Write(VInt elementId, ulong value, ulong defaultValue = 0)
```

**Comment by OlegZee** [comment]
Hm… The good point is that it’s backward compatible, i.e. 0 will be written as 80.Does it work well for your scenario: you will have to pass a different value in the last parameter? Also we can made Read methods similarly, i.e. `ReadInt(int defaultValue = 0)`. From: André KöhlerSent: 16 апреля 2021 г. 16:55To: OlegZee/NEbmlCc: Oleg; State changeSubject: Re: [OlegZee/NEbml] How to write "0" value? (#10) I suppose it's redundant.Which one would be used if I write code like this Write(DurationElementId, null)?Good point, I did not think about that. I just tested it with a test project and it seems like there is a priority between integer types when selecting between Nullable overloads. When having to chose between int?, uint?, long? and ulong? C# picks the overload with int? but as soon as an overload with string exists I get "error CS0121: The call is ambiguous". So my idea was not good I guess.Next idea: Specify the default value in the write call. If the value equals the default value, the element is written with a length of zero:public int Write(VInt elementId, long value, long defaultValue = 0)public int Write(VInt elementId, ulong value, ulong defaultValue = 0)—You are receiving this because you modified the open/close state.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by andrekoehler** [comment]
Passing the default value to each Write and Read method works for me, because my Matroska (de-/)serializer knows it for each element (I use a System.Attribute to store that information).
